File: leetcode/0517.number-of-provinces/go/001.union-find.go
func findCircleNum(isConnected [][]int) int {
    n := len(isConnected)
    provinces := n

    parent := make([]int, n)
    rank := make([]int, n)

    for i := 0; i < n; i++ {
        parent[i] = i
        rank[i] = 1
    }

    findRoot := func(node int) int {
        cur := node

        for cur != parent[cur] {
            parent[cur] = parent[parent[cur]]
            cur = parent[cur]
        }

        return cur
    }


    union := func(a, b int) bool {
        rootA := findRoot(a)
        rootB := findRoot(b)

        if rootA == rootB {
            return false
        }

        if rank[rootA] <= rank[rootB] {
            rank[rootB] += rank[rootA]
            parent[rootA] = rootB
        } else {
            rank[rootA] += rank[rootB]
            parent[rootB] = rootA
        }

        return true
    }


    for i := 0; i < n; i++ {
        for j := i+1; j < n; j++ {
            if isConnected[i][j] == 1 {
                if union(i, j)  {
                    provinces -= 1
                }
            }
        }
    }

    return provinces
}

File: leetcode/0517.number-of-provinces/js/001.union-find.ts
function findCircleNum(isConnected: number[][]): number {
    const n = isConnected.length;

    const rank = new Array(n).fill(1);
    const parent = Array.from({ length: n }, (_, i) => i);;

    let res = n;

    const findRoot = (node) => {
        let cur = node;

        while(cur !== parent[cur]) {
            parent[cur] = parent[parent[cur]];
            cur = parent[cur];
        }

        return cur;
    }

    const union = (a, b) => {
        const rootA = findRoot(a);
        const rootB = findRoot(b);

        if(rootA === rootB) {
            return false;
        }

        if(rank[rootA] <= rank[rootB]) {
            parent[rootA] = rootB;
            rank[rootB] += rank[rootA];
        } else {
            parent[rootB] = rootA;
            rank[rootA] += rank[rootB];
        }

        return true;
    }


      for(let i = 0; i < n; i++) {
        // Start from i+1 because matrix is symmetric and we don't need diagonal
        for(let j = i + 1; j < n; j++) {
            if(isConnected[i][j] === 1) {
                if(union(i, j)) {
                    res--; 
                }
            }
        }
    }

    return res;
};
